File: src/github/pr_metadata.py
import requests
import os
import logging
from .auth import get_github_app_token
from .pr_comments import get_pr_number_from_event

logger = logging.getLogger("agent-runner")

def fetch_pr_metadata():
    # Get GitHub App token
    github_token = get_github_app_token()
    if not github_token:
        logger.error("Failed to get GitHub App token")
        return None, None

    repo = os.getenv("GITHUB_REPOSITORY")
    pr_number = os.getenv("PR_NUMBER") or get_pr_number_from_event()

    if not github_token or not repo or not pr_number:
        logger.warning("Cannot fetch PR metadata: missing token, repo, or PR number")
        return None, None

    headers = {
        "Authorization": f"token {github_token}",
        "Accept": "application/vnd.github.v3+json",
    }

    url = f"https://api.github.com/repos/{repo}/pulls/{pr_number}"
    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        data = response.json()
        return data.get("title"), data.get("body")
    else:
        logger.error(f"Failed to fetch PR metadata: {response.status_code} {response.text}")
        return None, None

File: src/github/pr-metadata.ts
/**
 * GitHub PR metadata utilities.
 * Functions for fetching PR title and description from GitHub API.
 */

import { getGithubAppToken } from "./auth.js";
import { getPrNumberFromEvent } from "./pr-comments.js";

/**
 * PR metadata response from GitHub API.
 */
interface PrMetadata {
  title: string | null;
  description: string | null;
}

/**
 * GitHub PR API response structure.
 */
interface GitHubPrResponse {
  title?: string;
  body?: string;
}

/**
 * Fetch PR metadata (title and description) from GitHub API.
 * Equivalent to Python's fetch_pr_metadata().
 *
 * @returns Object with title and description, or null values if fetch fails
 */
export async function fetchPrMetadata(): Promise<PrMetadata> {
  // Get GitHub App token
  const githubToken = getGithubAppToken();
  if (!githubToken) {
    console.error("Failed to get GitHub App token");
    return { title: null, description: null };
  }

  const repo = process.env.GITHUB_REPOSITORY;
  const prNumberEnv = process.env.PR_NUMBER;
  let prNumber: number | null = null;

  if (prNumberEnv) {
    const parsed = parseInt(prNumberEnv, 10);
    prNumber = isNaN(parsed) ? null : parsed;
  } else {
    prNumber = await getPrNumberFromEvent();
  }

  if (!githubToken || !repo || !prNumber) {
    console.warn(
      "Cannot fetch PR metadata: missing token, repo, or PR number"
    );
    return { title: null, description: null };
  }

  const headers = {
    Authorization: `token ${githubToken}`,
    Accept: "application/vnd.github.v3+json",
  };

  const url = `https://api.github.com/repos/${repo}/pulls/${prNumber}`;

  try {
    const response = await fetch(url, { headers });

    if (response.status === 200) {
      const data = (await response.json()) as GitHubPrResponse;
      return {
        title: data.title || null,
        description: data.body || null,
      };
    } else {
      const errorText = await response.text();
      console.error(
        `Failed to fetch PR metadata: ${response.status} ${errorText}`
      );
      return { title: null, description: null };
    }
  } catch (error) {
    console.error("Error fetching PR metadata:", error);
    return { title: null, description: null };
  }
}

